src/Stratis.Bitcoin.Features.Dns/UnreliablePeerBehavior.cs

```diff
@@ -0,0 +1,136 @@
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public override object Clone()
+        {
+            return new UnreliablePeerBehavior(this.network, this.chainState, this.loggerFactory, this.peerBanning, this.nodeSettings);
+        }
+
+        /// <inheritdoc />
+        protected override void AttachCore()
+        {
+            this.AttachedPeer.MessageReceived.Register(this.OnMessageReceivedAsync);
+        }
+
+        /// <inheritdoc />
+        protected override void DetachCore()
+        {
+            this.AttachedPeer.MessageReceived.Unregister(this.OnMessageReceivedAsync);
+        }
+
+        /// <summary>
+        /// Event handler that is called when the node receives a network message from the attached peer.
+        /// </summary>
+        /// <param name="peer">Peer that sent us the message.</param>
+        /// <param name="message">Received message.</param>
+        /// <remarks>
+        /// This handler only cares about "version" messages, which are only sent once per node.
+        /// </remarks>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
+        {
+            if (message.Message.Payload is VersionPayload version)
+            {
+                // If current node is on POS, and ProvenHeaders is activated, check if current connected peer can serve Proven Headers.
+                // If it can't, disconnect from him and ban for few minutes
+                if (IsProvenHeaderActivated())
+                {
+                    if (!CanServeProvenHeader(version))
+                    {
+                        TimeSpan banDuration = TimeSpan.FromMinutes(1);
+                        this.logger.LogDebug("Peer '{0}' has been disconnected for {1} because can't serve proven headers. Peer Version: {2}", peer.RemoteSocketEndpoint, banDuration, version.Version);
+                        this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, "Can't serve proven headers.");
+                    }
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        /// Determines whether proven headers are activated based on the proven header activation height and applicable network.
+        /// </summary>
+        /// <returns>
+        /// <c>true</c> if proven header height is past the activation height for the corresponding network;
+        /// otherwise, <c>false</c>.
+        /// </returns>
+        private bool IsProvenHeaderActivated()
+        {
+            if (this.network.Consensus.Options is PosConsensusOptions options)
+            {
+                long currentHeight = this.chainState.ConsensusTip.Height;
+                return options.ProvenHeadersActivationHeight > 0 && currentHeight >= options.ProvenHeadersActivationHeight;
```

Comment:
brackets

Reply:
what do you mean? Should this be 
`return (options.ProvenHeadersActivationHeight > 0) && (currentHeight >= options.ProvenHeadersActivationHeight);` ?
or even more nested?

Reply:
yes, like you have it here